README.md
# The Glympse Adapter

- [Overview](#overview)
- [Client Mode](#client-mode)
- [Host Mode](#host-mode)
- [Examples](#examples)
- [Adapter Messages/Events](#adapter-messages/events)
- [Message/Event flow](#message/event-flow)
- [Adapter Endpoints](#adapter-endpoints)
- [Available modules](#available-modules)
- [Installation](#installation)
- [Glympse Adapter Roadmap](#glympse-adapter-roadmap)
- [Project Notes](#project-notes)
- [Release a new version](#release-a-new-version)

## Overview

The Glympse Adapter (GA) is a portable component that allows Javascript-based
applications to interface with the Glympse platform. This interface can be made
directly with client applications wishing to utilize additional components like the
Glympse viewer, or indirectly via an iframe interface in a pure data-only format.

The basic structure of the GA in action can be seen below:

![Glympse Adapter Overview](img/overview.png)

The GA automates many of the tasks involved with instantiating the Glympse viewer,
interfacing with the Glympse Platform API (from generating user accounts to requesting
data for a particular Glympse/Cards-based invite code), the Glympse Viewer API (adding/
removing invites, changing map settings, adjusting UI layouts, etc.), to invoking an
iframe-based view of a remote GA-enabled application, as shown below:

![Remote Adapter Integration](img/embed-scenario.png)

In addition, the API bridge provided by the GA maintains a mostly equivalent adaption
of the Glympse viewer API, save for some differences in event handler registration and
callback signatures. These (and all, really) will be described in detail. By default,
the exposed set of endpoints are those of the core Glympse Viewer API. Apps can also
extend/override the GA endpoints with their own methods, which are published to GA
consumers during setup.



## Client Mode
Below descibes how to utilize the GA in "client-mode", where full access is available
to all Viewer components, data structures, and many customization options. However, it
also requires more care in getting going, and maintaining a clean separation from the
Glympse Viewer, which will now be partially affected by the global namespace and
default styling options.

### Usage

The architecture of the GA running in client-mode is shown below:

![Glympse Adapter Client Mode](img/client-detail.png)

GA usage is straight-forward, but does have a dependency on a recent version of
[jQuery]. jQuery aside, all that is needed is to include the built
`GlympseAdapter` in your page, along with a bit of initialization and configuration
to get the adapter up and running:

```html
<script type="text/javascript" src="glympse-adapter-CURRENT_VERSION.min.js"></script>
```

Alternatively, the GA is available as a [Bower] package to include with your project
if using [RequireJS] via `bower install glympse-adapter --save`. If using the Bower
package, there is no need to include it as an external reference.

Initialization, including inserting the viewer into your page, should be done after
the page onLoad event has fired:

```javascript
$(window).ready(function()
{
	...
	... other init code
	...

	// Set up the adapter

	/* Global namespace */
	var adapter = new glympse.GlympseAdapter(adapter_handler_instance, cfg);

	/* AMD module format */
	//var GlympseAdapter = require('glympse-adapter/GlympseAdapter');
	//var adapter = new GlympseAdapter(adapter_handler_instance, cfg);

	adapter.client($('#div_viewer_container'));   // Note jQuery object reference
});
```

A couple of things to note here:
- `adapter_handler_instance` is a class instance to handle events and messages
  generated by the GA (described below).
- `cfg` is a simple Javascript object that defined both adapter and Glympse Viewer
  configuration (described below).
- `div_viewer_container` is the id of an HTML `div` element that already exists in
  the local DOM. The Glympse viewer will be inserted into this element when it is
  fully loaded.

For the `viewer_handler_reference` class, it must provide a public method with a
signature of `notify(idMessage, messageData)`. The `idMessage` parameter is a string
that will be one of the values as defined in the `glympse.GlympseAdapterDefines.MSG`
object. `messageData` can be a number of different types, depending on the value of
`idMessage`. Please refer to the *Adapter Messages/Events* section for specifics of
each message type.

GA client-mode  configuration format has two main components -- one for the Glympse
Viewer, and one for the adapter itself:

```javascript
var cfg =
{
	viewer: {
		.. normal viewer config settings ..
	}
	, adapter: {
		  card: *string*
		, t: *string*
		, pg: *string*
		, g: *string*
		, twt: *string*
		, hideEvents: *bool*
		, hideUpdates: *bool*
		, initialize: *function*
		, interfaces: { id0: callback0, ..., idN: callbackN }
		, apiKey: *string*|null|undefined
		, sandbox: *bool*
		, svcGlympse: *string*
		, dbg: *bool*
		, appName: *string*
		, appVersion: *string*
		, loaderEnvironment: *string*
		, loaderPath: *string*
		, loaderVersion: *string*
		, avatar: {
			minSize: *number* //in pixels
			, maxSize: *number* //in pixels
		}
	}
};
```

- `viewer` configuration is the normal Glympse viewer configuration object, described
  elsewhere. It should be noted that a valid invite/group/Twitter handler/Twitter
  topic setting should be set for normal viewer initialization.
- `adapter` settings are as follows:
  - `hideEvents`: Don't send event-related messages back to the
  `viewer_handler_reference.notify()` method. Currently, this includes everything
  except the `StateUpdate` message.
  - `hideUpdates`: Don't send `StateUpdate` messages back to the
  `viewer_handler_reference.notify()` method.
  - `hostQueueMaxSize`: [int] Max events to hold for adapter connecting in host-mode. Options:
    - 0: disable host message queue
    - < 0: no limit *(default)*
    - \> 0: limit to first `val` events
  - `initialize`: Method to call when the adapter has successully synced with a host
  adapater via iframe interface. Can be null.
  - `interfaces`: Allows for local overrides/extensions of advertised Glympse viewer
  endpoints. More details can be found in *Custom Endpoints*. Can also be null,
  - `apiKey`: Specifies the Core API key to use. If not set, the adapter will run in
  anonymous mode, which will allow only for the viewing of Core API invites.
  - `sandbox`: Force usage of the sandbox environment for Cards/Glympses. Note that
  this setting, if set, will change some defaults of other params to fully align to
  the sandbox environment: `loaderEnvironment`, `svcEnRoute` and `svcGlympse`
  - `svcEnRoute`: Override the default Glympse EnRoute API environment. Default is
    `//api.enterprice.glympse.com/api/v1/`, or `//s-api.enterprice.glympse.com/api/v1/`, if `sandbox` is set.
  - `svcGlympse`: Override the default Glympse API environment. Default is
  `//api.glympse.com/v2/`, or `//api.sandbox.glympse.com/v2/`, if `sandbox` is set.
  - `dbg`: Enable debugging state. Defaults to `false`.
  - `card`, `t`, `g`, `twt`, `pg`: Standard Cards/Glympse invite types.
  - `loaderEnvironment`: Environment for the viewer's loader: "prod"|"sandbox".
  Automatically set based on _sandbox_ setting if this param is not specified.
  - `loaderVersion`: Version of the viewer's loader. Default is "stable".
  - `loaderPath`: Arbitrary URL to the loader (overrides `loaderEnvironment` and `loaderVersion`). Default is NULL.
  - `avatar` - configuration for Image processing for `core.setUserAvatar` API endpoint
    - `minSize`: Min size for avatar image scaling
    - `maxSize`: Max size for avatar image scaling
  - `appName` - Name of the hosting application
  - `appVersion` - Version of the hosting application
  - `demoDriversCount` - The number of drivers to return in `demoshuttle` group. Min=0 and Max=7.
  - `demoWaypoints` - The list of waypoints which should be returned for the demo org object. 
  The format is `[ { name: 'Stop #1', lat: 123, lng: -123 }, ... ]`. The last point is base org location.

### Custom Marker Configuration
For the `addMarkers(cfgMarkers)` API, the `cfgMarkers` object is of the following
format:

	cfgMarkers = { showInfo: true|false, markers: [ marker_0_cfg, marker_1_cfg, ..., marker_N_cfg ] };

Each marker configuration is of the following format:

```javascript
var marker_N_cfg = { //ext: { path: image_url, id: null, w: width, h: height  }	// do not use if using icon setting
				 icon: { id: icon_id_from_spritesheet } // for internal
				, lat: latitude
				, lng: longitude
				, info: text_to_display // use \n to force a line break
				, track: true|false     // include for map-centering
				, link: url_to_load_if_marker_clicked
				, style: {
					  fontFamily: available_font_family
					, color: CSS_color
					, fontSize: font_size
					, maxWidth: max_width_in_pixels (forcing word wrap)
					, background: null|CSS_color
					, textMargin: margin_in_pixels
					}
				};
```

### Custom Route Options
For the `addRoute(cfgRoute)` API, the following configuration object is used:

```javascript
var cfgRoute = { start: [ lat, lng ]
			   , end: [ lat, lng ]
			   , useNubs: true|false // Add nubs to the start/end points
			   , nubSize: 6          // Nub radius
			   , nubStroke: 0.1      // Nub stroke
			   , style: { strokeColor: 'rgba(255,0,0,0.8)' }
		   };
```

### Custom Endpoints
While not usuaully necessary, the GA allows for extending and/or overriding the
default set of endpoints exposed by the Adapter if you need to allow for additional
functionality. As noted above, the format for specifying custom endpoints is in the
adapter's `interfaces` configuration setting:

```javascript
var cfg =
{
	viewer: {
		.. normal viewer config settings ..
	}
	, adapter: {
		.. other adapter settings ..
		, interfaces: { id0: callback0, ..., idN: callbackN }
	}
};
```

The `interfaces` object is an optional collection of methods to be made available for
GA consumers, or to override existing Glympse viewer methods. Each item in the
collection is decribed in the following manner:

- `id`: The name of the interface to advertise to GA host-mode consumers upon initial
adapter setup.
- `callback`: The method to execute when called from the GA host-mode consumer.

For `id`, if the name is the same as one of the pre-defined endpoints (listed above),
the default implementation is replaced with this custom version. However, the default
is still available via the id with the same name, but with a `base_` prefix (i.e.
`base_getValue` to reach the Glympse viewer's implementation).

Almost any valid Javascript name can be used, save for the default public method names
of the GA. To be safe, prefix the name of your custom endpoints with a unique name
(i.e. "my" would work wonderfully).

For now, your customized endpoints should return some value directly, if expected by
the caller. The current version of the GA does not support deferred responses. In the
next version, a Promise-based return value will allow for async handling for delayed
results (i.e. Ajax-based calls).

And for reference, the test app pair of `index-test.html` and `test/app.js` show an
example custom endpoint of `customMethodExample`. This can be easily extended and/or
enhanced for your needs.


## Host Mode
Below descibes how to utilize the GA in "host-mode", with limited access to Viewer
components and data structures. It runs in an iframe-based environment, providing a
clean separation to the rest of the page in which it is embedded. That said, the
communication channel established with the View-Mode-based GA control is fairly robust
and can handle a number of data types (including such things as image data).

### Usage

The architecture of the GA running in host-mode is shown below:

![Glympse Adapter Host Mode](img/host-detail.png)

Host-Mode usage has been designed to replicate the Glympse Viewer and Cards APIs,
adding a bit of initialization and configuration to get the adapter up and running.
All that is needed is to include the built GlympseAdapter.js in your page:

```html
<script type="text/javascript" src="glympse-adapter-VERSION.min.js"></script>
```

As noted in the Client Mode section, the GA can also be used in an AMD module
environment via Bower. Same set-up semantics apply when using GA in host mode.

Initialization, including inserting the viewer into your page, should be done after
the page onLoad event has fired:

```javascript
$(window).ready(function()
{
	...
	... other init code
	...

	// Set up the adapter

	/* Global namespace */
	var adapter = new glympse.GlympseAdapter(adapter_handler_instance, cfg);

	/* AMD module format */
	//var GlympseAdapter = require('glympse-adapter/GlympseAdapter');
	//var adapter = new GlympseAdapter(adapter_handler_instance, cfg);

	var el = adapter.host({
		url: 'http://glympse.com/INVITE_CODE',
		initialize: clientInitialized,
		connect: clientConnected,
		events:
		{
			StateUpdate: stateUpdate,
			ViewerInit: viewerInit,
			ViewerReady: viewerReady
		},
		sandbox:
		{
			disabled: true
		}
	});

	$('#divSandbox').append(el);

	...
});
```

A couple of things to note here:
- `adapter_handler_instance` is a class instance to handle events and messages
  generated by the GA (described below).
- `cfg` is a simple Javascript object that defined both adapter and Glympse Viewer
  configuration (described below).
- The iframed viewer is created/returned, when calling the GA instance's `host()`
  method.

`host()` configuration options are as follows:

- `url`: The Glympse invite URL. Generally this is the same as what is shared when
  sending a normal Glympse.
  - `partnerid`: Your app's registered PartnerId given to you by your
  Glympse representative
- `initialize`: Callback to handle iframe loaded event. Note that the Glympse viewer is
  not yet ready at this point, only that the page has been successfully loaded.
- `connect`: Callback to handle when the loaded Glympse viewer has registered itself
  with your hosting adapter. It is still not ready for action.
- `events`: Additional events to be handled by the host app. Not all events need to
  be handled if they aren't going to be used. Format is
  `{ event_id0: callback_handler0, ..., event_idN: callback_handlerN`. See the
  Host Events section below for available events that can be handled.
- `sandbox`: Configuration of sandbox environment:
    - `popups` `(default: false)` - allow opening windows from sandbox env
    - `disabled` `(default: false)` - (_only for iframe adapter_) do not create actual sandbox on iframe to allow all actions from iframed app
                   (**Note:** this is glympse specific feature, see notes [below](#project-notes))

All API calls to the client-mode adapter return a Promise, allowing for async handling
of all results. These should be treated like any baseline [Promise/A+] system:

```javascript
adapter.map.getInviteProperty({ idProperty: GlympseAdapterDefines.STATE.Name }).then(function(data)
{
	console.log('Glympse invite sender's name: ' + data);
}
```

A more detailed description of all of the available Glympse viewer APIs is described
elsewhere in this document. Additionally, refer to `app/test-host/index-host.html` as
a reference for using all available APIs with the GA running in host-mode.


### Host Event support
Events generated by the hosted client-based GA application can be hooked during host
mode initialization. All of the named events can be mapped by providing a function
callback. If no callback is provided, no event will be generated.

#### Host-specific Events
A couple low-level events have been added to monitor the loading and initialization
steps during the course of loading an iframe'd Glympse Adapter-based viewer. Both
of these settings are specified in the passed `host(cfg)` API configuration (outlined
elsewhere). However, these events are generally not needed for most applications,
except to monitor initialization and loading progress.

- `Initialize`: Invoked whenever the iframe html has been loaded and registered with
  the host container. Callback format is: `callback(name)`. The passed `name`
  parameter in the callback is the id used interface with the client-hosted viewer.

- `Connect`: Sent once the hosted GA application has loaded and initialized
  its interfaces with the host container. Callback format is: `callback(options)`.
  The passed `options` parameter in the callback shows the features of the hosted
  iframe'd client:
  - `invitesCard`: *(string)* List of Card invites to load
  - `invitesGlympse`: *(string)* List of Glympse invites to load (*string*)
  - `id`: *(string)* Connected client app identifier (generally "glympse-adapter")
  (*string*)
  - `version`: *(string)* Connected client GA version (*string*)
  - `map`: *(array)* Available map-based APIs
  - `cards`: *(array)* Available Glympse card-based APIs
  - `ext`: *(array)* Available custom APIs
  - `core`: *(array)* Available Glympse ticket-based APIs
  - `app`: *(array)* Available well-known application APIs



#### Cards/Map/Custom Events
All of the GA-based event handlers are specified in the `events` section of the
`host(cfg)` configuration (as outlined above). The names of the handlers are identical
to the name of the Adapter event to be handled. For example, to handle the `ViewerInit`
and `ViewerReady` events, the `host()` config object would look like:

```javascript
var htmlElement = adapter.host(
	{ url: ...,
	  events: { ViewerInit: myViewerInit, ViewerReady: myViewerReady }
	}
);
```
If you do not specify a handler for a particular GA event, it will not be raised by GA.


## Examples
### Client-mode
An example of setting up a GA client-mode-based application can be found in
`app/src/test-client`. This test app demonstrates various GA actions when a
Glympse ticket/card invite is loaded.

This example should be used as a reference of how to invoke and interact with
the GA in client-mode.

### Host-mode
An example of setting up a GA host-mode-based application can be found in
`app/src/test-host`. This test app demonstrates how to hook into a
remote GA client-mode-based application and leverage interesting events that
are generated when a Glympse ticket/card invite is loaded. There are also
examples of how to directly query the client-mode application to retrieve
specific data.

However, some setup is required to alias the local machine to point to the
`app/src/test-client/index-client.html` application. To use the example out of the box
(on an OSX-based system):

- Open a shell in the synced root directory of this repo
  - `cd app`
  - `python -m SimpleHTTPServer 8000`
- In another shell:
  - `cd app`
  - `python -m SimpleHTTPServer 8080`
- In a browser, connect to: `http://localhost:8080/src/test-host/index-host.html`

If the above isn't feasible for you, make whatever settings necessary to point to the
`app/src/test-client` directory to another port/domain and edit the `app.urlClient`
setting in `app/src/test-host/index-host.html` (on or around line 42).

This example should be used as a reference of how to invoke and interact with the GA
in host-mode.

### Simultaneous Host/Client mode
In some scenarios, it is useful to have an app utilizing the GA in both
host and client modes. In such cases, multiple instances of the adapter need
to be created to support this scenario, as the adapter is currently locked
to a single mode once it is initialized.

A simple example that leverages the code from the above host and client
examples can be found in `app/src/test-host/client`. Check out `src/app.js`
as the reference for instantiating two GAs simultaneously.

As with the *Host mode* example, the same setup is required for running the
host-mode portion of the demo, changing the path to load:
`http://localhost:8080/src/test-host-client/index-host-client.html`


## Adapter Messages/Events

The following is a description of the various notification messages (or Events, if
using GA in host-mode) that are sent by the adapter (defined in the
`GlympseAdapterDefines.MSG` object):

- `AccountCreateStatus` / `{ status: bool }`: Indicates the account has been created or
  creation failed with `error` message and `status: false`
- `AccountLoginStatus` / `{ status: bool }`: `status: true` - Indicates the adapter successfully
  logged in. `status: false` - Error occurred, property `error` has additional info about it
- `AccountDeleteStatus` / `{ status: bool }`: `status: true` - Indicates the adapter successfully
  removed current credentials.
- `AdapterInit` / `{ isCard: bool, t: string }`: Indicates the adapter is beginning
  its loading sequence with the passed card invite or the specified Glympse invite
  (indicated via the `t` parameter)
- `AdapterReady` / `{ cards:[ invites ], glympses: [ invites ]}`: Specifies the
  adapter has resolved all Card and Glympse invites and will begin the process
  of loading the specified Glympse invites. It should be noted that the adapter
  may return a card that was not specified in the `AdapterInit` phase, and may
  have changed the initial Glympse invite code, or included additional Glympse
  invites. This is due to the adapter detecting a Card invite reference in the
  initial Glympse invite code, and aligning the viewer to the Card reference and
  it's child Glympse invites.
- `CreateRequestStatus` / `{ status: bool }`: Indicates the request was created or 
  creation failed with `error` message and `status: false`
- `CardInit` / `card_invite`: Fired just before an attempt is made to load the
  Card invite.
- `CardReady` / `card_invite`: Fired once the given Card invite has completed
  loading, regardless of success. Check the passed card's instance given upon
  the `CardsInitEnd` event.
- `CardsInitEnd` / `[ card_instance0, ..., card_instanceN ]`: Sent once all
  cards have been loaded. The passed array are a set of Card.js class instances
  that can be check for various properties/state. Please reference the `Card.js`
  Data Model section for available APIs.
  - Note that in GA Host mode, Card instances are not passed
- `CardsInitStart` / `[ card_invite0, ..., card_inviteN ]`: Notification sent
  just before Card data is loaded. The passed array is a list of the Card invite
  codes that will be loaded.
- `CardAdded` / `card`: Notification sent when new card is added and loaded.
- `CardRemoved` / `card`: Notification sent when card is removed.
- `CardUpdated`: Notification sent when card is updated (e.g. members' list changed).
  Event model:
  - `card: Card`: model of changed card
  - `action: string`: possible actions:
    - `member_added`
    - `member_removed`
    - `invite_added`
    - `invite_removed`
    - `member_started_sharing`
    - `member_stopped_sharing`
    - other actions described in the [spec](https://developer.glympse.com/docs/core/api/reference/cards/id/activity/get).
  - `member: Member`: model of card member (only for `member_*` events)
  - `invite: string`: invite code (only for invite's related events)
  - `data: Object`: not parsed response for all other actions
  
- `CardsJoinRequestStatus`: Indicates that Card join request has been created or
  creation failed with `response.error` message and `status: false`
- `CardsJoinRequestCancelStatus`: Indicates that Card join request has been cancelled or
  cancelling failed with `response.error` message and `status: false`
- `CardActivityStatus`: Returns the list of card activities for the requested period
- `CardsActiveJoinRequestsStatus`: Returns the list of active join card requests in `response` property
- `CardRemoveMemberStatus`: Indicates that Card member was removed or removing failed with `response.error` message and `status: false`
- `CardsLocationRequestStatus`: Indicates that location request was created or creation failed with `response.error` message and `status: false`
- `DataUpdate` / `{ id: glympse_invite_code, owner: glympse_user_account_id, card: card_id, data: [ property_0, ..., propertyM ] }`:
  Event passed from the Glympse API for a given Glympse invite code, for unknown/custom
  properties found in the Glympse's data stream. The format of property elements
  in the array are the same format as specified in the Glympse Data stream model:
  - `t`: Time property was generated
  - `n`: Property id
  - `v`: Property value (may be default type, or a custom Object with additional
  members)
- `EtaInfo`: Generates when ETA info was fetched in response to `getEtaInfo` request to the `core` endpoint, 
  with information the loaded ETAs. The order in array the same as in request args.
- `GroupLoaded`: Generated after an `addGroup` command is sent to the `core` endpoint, with information
  about the just-loaded Glympse public group. For normal responses, the `args` payload will look something
  like:
  ```
  {
    "status": true,
    "response": {
      "type": "group",
      "id": 119,
      "events": 1,
      "members": [
        {
          "id": "DNA7-4HDZ-03WH0",
          "invite": "demobot0"
        }
      ],
      "public": true,
      "name": "DemoShuttle"
    },
    "time": 1499192505285,
    "group": "demoshuttle",
    "invitesAdded": [
      "demobot0"
    ],
    "invitesRemoved": [],
    "invitesSwapped": []
  }
  ```
- `GroupStatus`: Seen periodically after a public group has been added to the adapter. This event
  will contain updates to the group after initial load. Main usage will be in examining the contents
  of the `invitesAdded`, `invitesRemoved`, and `invitesSwapped` arrays, which will be the updates since
  initial load. Note that this event data will be identical to the initial `GroupLoaded` event for the
  first `GroupResponse` event seen. See `GroupLoaded` for expected data format.
- `OrgObjects`: Generated after an `getOrgObjects` command is sent to the `core` endpoint, with information

CHANGELOG.md
#[Changelog] glympse-adapter

version |date        |notes
:-------|:-----------|:------
2.13.1  | 2018/08/16 | Added `grunt jsdoc` task + some jsdoc annotations 
2.13.0  | 2018/06/29 | Add interface to access eta_range property, remove as data stream property
2.12.2  | 2018/06/21 | Added eta_range data stream property
2.12.1  | 2018/05/17 | Added ability to skip "sandbox" using on iframe (host-mode)
2.12.0  | 2018/05/15 | Added GlympseAdapter.app.sendOasisMessage API (for client mode only)
2.11.1  | 2017/12/19 | Bugfix: de-duplicate invite's data-stream.
2.11.0  | 2017/10/06 | Added waypoints support for `core.getEtaInfo` API. 
2.10.0  | 2017/09/08 | BugFix for handling public group events (cleanup the events queue before processing).
2.9.0   | 2017/08/01 | BugFix for auth requests. New `core.getEtaInfo` endpoint + `EtaInfo` event.
2.8.1   | 2017/07/19 | New `ViewerZoomChanged` event + bot group options
2.7.2   | 2017/07/17 | Missing auth / removed ticket invite bugfixes
2.7.1   | 2017/07/12 | Propagate app info through to viewer
2.7.0   | 2017/07/12 | Add removeGroup and getGroups APIs for additional handling of Glympse public groups + various bugfixes.
2.6.0   | 2017/07/05 | Add support for loading Glympse public groups directly in the adapter.
Client must currently manage invites (i.e. manually add/remove them on the map -- see examples)
2.5.0   | 2017/06/30 | Update all api calls to send adapter/hosting app name and version
2.4.0   | 2017/06/17 | New events, better token caching, https consistency, updated examples, new app interface
2.3.2   | 2017/04/27 | Fix for removing first ticket invite from card
2.3.1   | 2017/04/06 | Additional handling for expired/invalided auth tokens
2.3.0   | 2017/03/10 | `hostQueueMaxSize` option: 0 = disable host message queue, <0 = no limit, >0 = limit to first N events
2.2.0   | 2017/03/07 | Change default viewer version to "stable"
2.1.2   | 2017/03/07 | Bug fix for ticket-only events
2.1.0/1 | 2017/02/26 | Cleaned up support for `sandbox=1` setting. Removed `svcCards` setting (only `svcGlympse` is necessary now). Point to correct viewer base URL.
2.0.2	| 2017/02/17 | Add `invitee` property on the Card model for proper handling of *card_request* invite types
2.0.1	| 2017/02/15 | Updated to use LocalStorage instead of cookies, if available
2.0.0   | 2017/02/10 | Revamp of internal architecture + addition of Glympse Cards APIs. Biggest imact for ticket invite usage: must specify apiKey param and set isAnon=true config setting for regular viewing of ticket invites.
1.4.2   | 2016/11/12 | *InviteClicked* event propagation fix
1.4.1   | 2016/11/09 | Add *InviteClicked* event
1.4.0   | 2016/11/08 | Remove `adapter.getViewer()`, add client-only `adapter.map.getMapContainer()` method
1.3.10  | 2016/07/23 | Expire option for `setCfgVal()`: `lib.setCfgVal(name, val, daysExpire)` _(default: 365)_
...     | ...        | **\[ _Need to pull in from tagged version release notes, as available_ \]**
1.0.0   | 2015/08/11 | Initial release


app/src/lib/md5.js
/*
 * JavaScript MD5
 *
 * IMPORTANT:
 *  - Based on https://github.com/blueimp/JavaScript-MD5 - v2.7.0
 *  - Contains small tweaks + removed unnecessary methods
 */

/* global define */

(function(w)
{
	'use strict';

	/*
	 * Add integers, wrapping at 2^32. This uses 16-bit operations internally
	 * to work around bugs in some JS interpreters.
	 */
	function safeAdd(x, y)
	{
		var lsw = (x & 0xFFFF) + (y & 0xFFFF);
		var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
		return (msw << 16) | (lsw & 0xFFFF);
	}

	/*
	 * Bitwise rotate a 32-bit number to the left.
	 */
	function bitRotateLeft(num, cnt)
	{
		return (num << cnt) | (num >>> (32 - cnt));
	}

	/*
	 * These functions implement the four basic operations the algorithm uses.
	 */
	function md5cmn(q, a, b, x, s, t)
	{
		return safeAdd(bitRotateLeft(safeAdd(safeAdd(a, q), safeAdd(x, t)), s), b);
	}

	function md5ff(a, b, c, d, x, s, t)
	{
		return md5cmn((b & c) | ((~b) & d), a, b, x, s, t);
	}

	function md5gg(a, b, c, d, x, s, t)
	{
		return md5cmn((b & d) | (c & (~d)), a, b, x, s, t);
	}

	function md5hh(a, b, c, d, x, s, t)
	{
		return md5cmn(b ^ c ^ d, a, b, x, s, t);
	}

	function md5ii(a, b, c, d, x, s, t)
	{
		return md5cmn(c ^ (b | (~d)), a, b, x, s, t);
	}

	/*
	 * Calculate the MD5 of an array of little-endian words, and a bit length.
	 */
	function binlMD5(x, len)
	{
		/* append padding */
		x[len >> 5] |= 0x80 << (len % 32);
		x[(((len + 64) >>> 9) << 4) + 14] = len;

		var i;
		var olda;
		var oldb;
		var oldc;
		var oldd;
		var a = 1732584193;
		var b = -271733879;
		var c = -1732584194;
		var d = 271733878;

		for (i = 0; i < x.length; i += 16)
		{
			olda = a;
			oldb = b;
			oldc = c;
			oldd = d;

			a = md5ff(a, b, c, d, x[i], 7, -680876936);
			d = md5ff(d, a, b, c, x[i + 1], 12, -389564586);
			c = md5ff(c, d, a, b, x[i + 2], 17, 606105819);
			b = md5ff(b, c, d, a, x[i + 3], 22, -1044525330);
			a = md5ff(a, b, c, d, x[i + 4], 7, -176418897);
			d = md5ff(d, a, b, c, x[i + 5], 12, 1200080426);
			c = md5ff(c, d, a, b, x[i + 6], 17, -1473231341);
			b = md5ff(b, c, d, a, x[i + 7], 22, -45705983);
			a = md5ff(a, b, c, d, x[i + 8], 7, 1770035416);
			d = md5ff(d, a, b, c, x[i + 9], 12, -1958414417);
			c = md5ff(c, d, a, b, x[i + 10], 17, -42063);
			b = md5ff(b, c, d, a, x[i + 11], 22, -1990404162);
			a = md5ff(a, b, c, d, x[i + 12], 7, 1804603682);
			d = md5ff(d, a, b, c, x[i + 13], 12, -40341101);
			c = md5ff(c, d, a, b, x[i + 14], 17, -1502002290);
			b = md5ff(b, c, d, a, x[i + 15], 22, 1236535329);

			a = md5gg(a, b, c, d, x[i + 1], 5, -165796510);
			d = md5gg(d, a, b, c, x[i + 6], 9, -1069501632);
			c = md5gg(c, d, a, b, x[i + 11], 14, 643717713);
			b = md5gg(b, c, d, a, x[i], 20, -373897302);
			a = md5gg(a, b, c, d, x[i + 5], 5, -701558691);
			d = md5gg(d, a, b, c, x[i + 10], 9, 38016083);
			c = md5gg(c, d, a, b, x[i + 15], 14, -660478335);
			b = md5gg(b, c, d, a, x[i + 4], 20, -405537848);
			a = md5gg(a, b, c, d, x[i + 9], 5, 568446438);
			d = md5gg(d, a, b, c, x[i + 14], 9, -1019803690);
			c = md5gg(c, d, a, b, x[i + 3], 14, -187363961);
			b = md5gg(b, c, d, a, x[i + 8], 20, 1163531501);
			a = md5gg(a, b, c, d, x[i + 13], 5, -1444681467);
			d = md5gg(d, a, b, c, x[i + 2], 9, -51403784);
			c = md5gg(c, d, a, b, x[i + 7], 14, 1735328473);
			b = md5gg(b, c, d, a, x[i + 12], 20, -1926607734);

			a = md5hh(a, b, c, d, x[i + 5], 4, -378558);
			d = md5hh(d, a, b, c, x[i + 8], 11, -2022574463);
			c = md5hh(c, d, a, b, x[i + 11], 16, 1839030562);
			b = md5hh(b, c, d, a, x[i + 14], 23, -35309556);
			a = md5hh(a, b, c, d, x[i + 1], 4, -1530992060);
			d = md5hh(d, a, b, c, x[i + 4], 11, 1272893353);
			c = md5hh(c, d, a, b, x[i + 7], 16, -155497632);
			b = md5hh(b, c, d, a, x[i + 10], 23, -1094730640);
			a = md5hh(a, b, c, d, x[i + 13], 4, 681279174);
			d = md5hh(d, a, b, c, x[i], 11, -358537222);
			c = md5hh(c, d, a, b, x[i + 3], 16, -722521979);
			b = md5hh(b, c, d, a, x[i + 6], 23, 76029189);
			a = md5hh(a, b, c, d, x[i + 9], 4, -640364487);
			d = md5hh(d, a, b, c, x[i + 12], 11, -421815835);
			c = md5hh(c, d, a, b, x[i + 15], 16, 530742520);
			b = md5hh(b, c, d, a, x[i + 2], 23, -995338651);

			a = md5ii(a, b, c, d, x[i], 6, -198630844);
			d = md5ii(d, a, b, c, x[i + 7], 10, 1126891415);
			c = md5ii(c, d, a, b, x[i + 14], 15, -1416354905);
			b = md5ii(b, c, d, a, x[i + 5], 21, -57434055);
			a = md5ii(a, b, c, d, x[i + 12], 6, 1700485571);
			d = md5ii(d, a, b, c, x[i + 3], 10, -1894986606);
			c = md5ii(c, d, a, b, x[i + 10], 15, -1051523);
			b = md5ii(b, c, d, a, x[i + 1], 21, -2054922799);
			a = md5ii(a, b, c, d, x[i + 8], 6, 1873313359);
			d = md5ii(d, a, b, c, x[i + 15], 10, -30611744);
			c = md5ii(c, d, a, b, x[i + 6], 15, -1560198380);
			b = md5ii(b, c, d, a, x[i + 13], 21, 1309151649);
			a = md5ii(a, b, c, d, x[i + 4], 6, -145523070);
			d = md5ii(d, a, b, c, x[i + 11], 10, -1120210379);
			c = md5ii(c, d, a, b, x[i + 2], 15, 718787259);
			b = md5ii(b, c, d, a, x[i + 9], 21, -343485551);

			a = safeAdd(a, olda);
			b = safeAdd(b, oldb);
			c = safeAdd(c, oldc);
			d = safeAdd(d, oldd);
		}
		return [a, b, c, d];
	}

	/*
	 * Convert an array of little-endian words to a string
	 */
	function binl2rstr(input)
	{
		var i;
		var output = '';
		var length32 = input.length * 32;
		for (i = 0; i < length32; i += 8)
		{
			output += String.fromCharCode((input[i >> 5] >>> (i % 32)) & 0xFF);
		}
		return output;
	}

	/*
	 * Convert a raw string to an array of little-endian words
	 * Characters >255 have their high-byte silently ignored.
	 */
	function rstr2binl(input)
	{
		var i;
		var output = [];
		output[(input.length >> 2) - 1] = undefined;
		for (i = 0; i < output.length; i += 1)
		{
			output[i] = 0;
		}
		var length8 = input.length * 8;
		for (i = 0; i < length8; i += 8)
		{
			output[i >> 5] |= (input.charCodeAt(i / 8) & 0xFF) << (i % 32);
		}
		return output;
	}

	/*
	 * Calculate the MD5 of a raw string
	 */
	function rstrMD5(s)
	{
		return binl2rstr(binlMD5(rstr2binl(s), s.length * 8));
	}

	/*
	 * Convert a raw string to a hex string
	 */
	function rstr2hex(input)
	{
		var hexTab = '0123456789abcdef';
		var output = '';
		var x;
		var i;
		for (i = 0; i < input.length; i += 1)
		{
			x = input.charCodeAt(i);
			output += hexTab.charAt((x >>> 4) & 0x0F) + hexTab.charAt(x & 0x0F);
		}
		return output;
	}

	/*
	 * Encode a string as utf-8
	 */
	function str2rstrUTF8(input)
	{
		return unescape(encodeURIComponent(input));
	}

	/*
	 * Take string arguments and return either raw or hex encoded strings
	 */
	function rawMD5(s)
	{
		return rstrMD5(str2rstrUTF8(s));
	}

	function hexMD5(s)
	{
		return rstr2hex(rawMD5(s));
	}

	function md5(string)
	{
		return hexMD5(string);
	}

	if (typeof define === 'function' && define.amd)
	{
		define(function()
		{
			return md5;
		});
	}
	else if (typeof module === 'object' && module.exports)
	{
		module.exports = md5;
	}
	else
	{
		// Global namespace registration
		var g = w.glympse || (w.glympse = {});
		var lib = g.lib || (g.lib = {});
		lib.md5 = md5;
	}

}(window));
